src/datastructures.jl

```diff
@@ -186,6 +251,38 @@ function Base.show(io::IO, gui::GUI)
     return dump(io, gui; maxdepth=1)
 end
 
+"""
+    iterate(iter::EnergySystemIterator)
+
+Initialize the iteration over an `EnergySystemIterator`, returning the first `EnergySystemDesign` object
+in the stack and the iterator itself. If the stack is empty, return `nothing`.
+"""
+function Base.iterate(iter::EnergySystemIterator)
+    isempty(iter.stack) && return nothing
+    current = pop!(iter.stack)
+    if isa(current, EnergySystemDesign)
+        push!(iter.stack, current.components...)  # Add the components to the stack
+        push!(iter.stack, current.connections...)  # Add the connections to the stack
+    end
+    return current, iter
+end
+
+"""
+    iterate(design::EnergySystemDesign)
+
+Initialize the iteration over an `EnergySystemDesign`, returning the first `EnergySystemDesign` object
+and the iterator itself. The iteration traverses through all nested components and connections.
+"""
+Base.iterate(design::EnergySystemDesign) = iterate(EnergySystemIterator([design]))
+
+"""
+    iterate(design::EnergySystemDesign, state)
+
+Initialize the iteration over an `EnergySystemDesign`, returning the first `EnergySystemDesign` object
+and the iterator itself. The iteration traverses through all nested components and connections.
+"""
+Base.iterate(::EnergySystemDesign, state) = iterate(state)
```

Comment:
I am not entirely certain whether the iterator represents the classical concept of an iterator, _i.e._, go through the elements of a type based on predefined size. Currently, it looks to me as if you in the first step create everything necessary. I would suggest instead to do that in the construction of the iterator type.